### DISCUSSION TITLE
Question about LastNeighborLoader

### DISCUSSION THREAD

**Discussion by Bowen-n** [pyg-team/pytorch_geometric#2153] [Q&A] [answered]
I'm studying TGN recently and reading [example code](https://github.com/rusty1s/pytorch_geometric/blob/master/examples/tgn.py). I don't understand what [`LastNeighborLoader`](https://github.com/rusty1s/pytorch_geometric/blob/6426b92141ab2fcb3d2b7b4c90074a63af725842/torch_geometric/nn/models/tgn.py#L210) is doing.

It contains data `neighbors`, `e_id`, `__assoc__`, what do these variables mean?

**Comment by rusty1s** [comment, marked as answer]
`LastNeighborLoader` will return the last encountered neighbohrs of `n_id` in an ever-growing graph. The idea is to insert new temporal edges via `insert`, and sample encountered edges via `__call__`.

Here, `e_id` is used as a reference to the global index of the returned edge. This helps to gather the respective edge features from your dataset belonging to the sampled edges.
`__assoc__` is used to map global edge indices to the local ones belonging to the sampled subgraph. For example, your original graph may contain 10,000 nodes, but your sampled subgraph does only reference 100 of them.

**Reply by Bowen-n** [reply]
Thanks.
TGN example code uses JODIE wikipedia dataset and it contains edge features but no node feature.
How can I use TGN to deal with temporal dataset that contains both node and edge features?

**Reply by rusty1s** [reply]
I guess you can simple concatenate node features and memory information, before inputting them into the GNN, e.g., [here](https://github.com/rusty1s/pytorch_geometric/blob/master/examples/tgn.py#L121).

**Reply by Bowen-n** [reply]
Thanks. I'll try it :)

**Reply by Bowen-n** [reply]
```python
n_id, edge_index, e_id = neighbor_loader(n_id)
...
z = gnn(z, last_update, edge_index, data.t[e_id], data.msg[e_id])
```
I still have some questions about `LastNeighborLoader`.
What does `edge_index` that `LastNeighborLoader` returns mean? And why it should be fed into `gnn`?

**Reply by rusty1s** [reply]
The `edge_index` will return a subgraph of all encountered edges, which will contain the last x interactions to other nodes from the nodes in `n_id`.

**Reply by Bowen-n** [reply]
It seems that current research on temporal graph networks are all applied on tasks like edge prediction or node classification. Is there any approach to apply temporal graph networks on graph-level task like graph classification?
For example, I have a temporal graph dataset. Each graph has a class label. And each edge in the graph has a timestamp representing the temporal information. And I want to conduct graph classification on this dataset.
How can I apply TGN on this task? In TGN, batches are formed from a single graph but in a graph classification task batches are formed from the whole graph dataset.
Is it possible to simply regard this dataset as a static graph dataset and just encode the timestamp to a kind of edge feature?

These questions are not about PyG. I'll be very glad if you could help. 😊

**Reply by rusty1s** [reply]
Indeed, there does not seem to be much research regarding temporal graph networks for graph-level tasks. I feel that `TGN` is not the right use-case here, and you may have multiple options here:
1. Regard this dataset as a static graph. This should already work just fine.
2. Create multiple graph instances w.r.t to different timestamps. You can utilize GNNs to obtain graph-level predictions on each of those graphs in isoloation, and input them into a GRU/LSTM afterwards.

**Reply by Bowen-n** [reply]
Thanks for your advice :)
